### PR TITLE
BeamSearch op cuda

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: 
Add CUDA implementation of BeamSearch operator.
(1) Support float16
(2) Implement all logits processors

The following are not included but might be available in separated pull request:
* Allocate ping-pong buffer for past state so that memory allocation is done once.
* Move initial feeds to CUDA kernel.
* Move update feeds to CUDA kernel.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Improve performance of text generation with beam search.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 1]
```diff
@@ -77,91 +83,144 @@ struct BeamSearchState {
...
-    next_positions = AllocateBuffer<int64_t>(allocator, next_positions_buffer_, batch_beam_size, true, static_cast<int64_t>(0));
+    this->next_positions = AllocateBuffer<int32_t>(allocator, next_positions_buffer_, batch_beam_size);
+
+    this->beam_scores = AllocateBuffer<float>(allocator, beam_scores_buffer_, batch_beam_size);
 
     if (output_scores) {
       size_t elements = SafeInt<size_t>(max_length - sequence_length) * batch_size * num_beams * vocab_size;
-      scores = AllocateBuffer<T>(allocator, scores_buffer_, elements);
-      remaining_scores = scores;
+      this->scores = AllocateBuffer<float>(allocator, scores_buffer_, elements);
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:101]: Do we keep the scores float when T=MLFloat16?

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+      sequences,
+      next_scores,
+      next_tokens,
+      next_indices);
+
+  return Status::OK();
+}
+
+template <typename T>
+Status DeviceCopy(gsl::span<T> target, gsl::span<const T> source, void* /*stream*/, int /*copyDirection*/) {
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:329]: I might miss sth but is there any implementation of DeviceCopy for H2D and D2H?

Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:329]: never mind. saw it under /cuda

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.h [hunk 3]
```diff
@@ -2,24 +2,27 @@
...
 
-template <typename T>
-class BeamSearch : public controlflow::IControlFlowKernel {
+using namespace onnxruntime::controlflow;  // namespace of IControlFlowKernel
+
+class BeamSearch : public IControlFlowKernel {
  public:
-  BeamSearch(const OpKernelInfo& info) : IControlFlowKernel(info) { Init(info); }
+  BeamSearch(const OpKernelInfo& info) : IControlFlowKernel(info), cuda_stream_(nullptr), dumper_(nullptr) {
+    Init(info);
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.h:23]: curious why there is a separate Init() in ctor?

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 4]
```diff
@@ -77,91 +83,144 @@ struct BeamSearchState {
...
 };
 
+struct BeamSearchCpuState : public IBeamSearchCpuState {
+  Sequences sequences;
+
+  void Init(AllocatorPtr allocator, size_t batch_beam_size, int max_length, bool is_cuda) {
+    this->sequence_lengths = AllocateBuffer<int32_t>(allocator, sequence_lengths_buffer_, batch_beam_size);
+    this->sequences_space = AllocateBuffer<int32_t>(allocator, sequences_space_buffer_, SafeInt<size_t>(2) * batch_beam_size * max_length);
+
+    if (is_cuda) {
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:123]: Is it a bit confused that CUDA buffers occur in a class named BeamSearchCpuState?

File: onnxruntime/contrib_ops/cuda/transformers/beam_search.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  SetDeviceHelpers(BeamSearchCudaDeviceHelper::ProcessLogits<MLFloat16>,
+                   BeamSearchCudaDeviceHelper::InitBeamState<MLFloat16>,
+                   BeamSearchCudaDeviceHelper::UpdateFeeds<MLFloat16>);
+
+  SetConsoleDumper(&g_cuda_dumper);
+}
+
+Status BeamSearch::ComputeInternal(OpKernelContext* context) const {
+  return onnxruntime::contrib::transformers::BeamSearch::Compute(context);
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/transformers/beam_search.cc:56]: So it calls cpu kernel and part of the impls are on cuda controlled by devicehelper. May I know the benefit of design this way?

**Comment by wangyems** [comment]
Didn't review the logic of beam search as it shares the same tests with cpu kernel.
lgtm if conflicts get resolved.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.h (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.h:23]: Init() has some initialization code. It is possible to merge code of Init() to ctor.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc (hunk 4 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:123]: BeamSearchCpuState has buffers allocated in CPU. Current CUDA implementation need copy some tensors to CPU for some processing.

File: onnxruntime/contrib_ops/cuda/transformers/beam_search.cc (hunk 5 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/transformers/beam_search.cc:56]: For beam search, some code (like subgraph execution) can be shared between CPU and CUDA. Device helpers are device specified implementation. It is quite similar to virtual function but better in isolating CPU and CUDA implementations.

**Review by wangyems — APPROVED** [review]
